### PR TITLE
[Merged by Bors] - refactor: use `FunLike` for `OuterMeasure`

### PR DESCRIPTION
**Pull request by urkud**
Also extracts `trim_anti_measurableSpace`

---

The scope of this PR is intentionally limited so that the PR is small. I'm going to migrate measures, then add more classes in subsequent PRs.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
Note that this is now the third time this has been attempted, #8453 and #8375. Both are stale of course, so maybe this one can replace them.

**Comment by urkud** [comment]
This one already passes the CI. I suggest that we merge it, then we can look at the old ones to see if there are parts that are still relevant.

File: Mathlib/MeasureTheory/Constructions/Pi.lean [hunk 1]
```diff
@@ -250,7 +250,7 @@ theorem tprod_tprod (l : List δ) (μ : ∀ i, Measure (π i)) [∀ i, SigmaFini
     (s : ∀ i, Set (π i)) :
     Measure.tprod l μ (Set.tprod l s) = (l.map fun i => (μ i) (s i)).prod := by
   induction' l with i l ih; · simp
-  rw [tprod_cons, Set.tprod, prod_prod, map_cons, prod_cons, ih]
+  erw [tprod_cons, Set.tprod, prod_prod, map_cons, prod_cons, ih]
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Constructions/Pi.lean:253]: I think #12686 will make this `erw` unnecessary

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean [hunk 2]
```diff
@@ -85,7 +85,7 @@ lemma MeasureTheory.QuotientMeasureEqMeasurePreimage.smulInvariantMeasure_quotie
     obtain ⟨𝓕, h𝓕⟩ := hasFun.ExistsIsFundamentalDomain
     have h𝓕_translate_fundom : IsFundamentalDomain Γ.op (g • 𝓕) ν :=
       h𝓕.smul_of_comm g
-    rw [h𝓕.projection_respects_measure_apply (μ := μ)
+    erw [h𝓕.projection_respects_measure_apply (μ := μ)
       (meas_π (measurableSet_preimage (measurable_const_smul g) hA)),
       h𝓕_translate_fundom.projection_respects_measure_apply (μ := μ) hA]
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:90]: If only one of these needs an `erw`, can you separate it into a `rw` and an `erw`? Same on the previous one.

**Review by eric-wieser — APPROVED** [review]
bors d+

Let's make the scope of the `erw` a little more fine-grained, otherwise this looks fine to go in.

**Comment by eric-wieser** [comment]
!bench

**Comment by urkud** [comment]
I'm benchmarking the next PR as well to see if some of these issues go away once we use `FunLike` for measures as well.

**Comment by urkud** [comment]
@eric-wieser Do you know how to figure out which theorems became slower?

**Comment by urkud** [comment]
!bench

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**